### ISSUE TITLE
Submodules clock and reset signals are not connected when using Circuit.Hierarchy

### ISSUE THREAD

**Issue by xguerin** [ujamjar/hardcaml#18]
The title says it all :) Tiny example:

```ocaml
    my_module
        the_my_module
        ( .dat_out(_1782[78:15]), .mdl_out(_1782[14:12]), .ctl_out(_1782[11:4]), .err_out(_1782[3:3]), .eob_out(_1782[2:2]), .sob_out(_1782[1:1]), .vld_out(_1782[0:0]) );
```

**Comment by andrewray** [comment]
You will need to explicitly wire up the clock.  That is include it in sub-module interfaces, and pass it into the reg etc functions.  The `Make_seq` functor, with a bit of first class module magic, could help ie something like 

```
module Sys = struct
  type 'a t = { clk : 'a; reset : 'a }[@@deriving hardcaml]
end

let seq sys = 
  let module X = Make_seq(struct 
    let reg_spec = { Seq.r_async with reg_clock = sys.Sys.clk; reg_reset = sys.Sys.rst } 
    let ram_spec = Seq.r_none end) 
  in
 (module X : Seq);;
```

Now if you pass Sys as a nested module in your interfaces you can do something like:

```
let f inputs = 
  let module Seq = (val (seq inputs.sys)) in
  Seq.reg ~e:vdd (inputs.a +: inputs.b);;
```

Unfortunately, you have to pass a Sys.t arg and add the `let moule Seq ...` line in all the functions that need register/memories, but I think it helps to package the problem up somewhat.

**Comment by xguerin** [comment]
Thanks for the explanation. It's a bit verbose, but that actually makes sense as I believe it enable the use of multiple clock domains.

**Comment by andrewray** [comment]
Yes - I think the general case of multiple clock domains would need this level of detail to control the wiring appropriately.

That said I wouldn't be adverse to a bodge when the design was just using the default global clock/reset/clear/enable stuff and get that wired up automatically.  It's been done in a few other places.

**Comment by andrewray** [comment]
I just checked in a fix to the hierarchy stuff related to sharing sub-circuits that are 'structurally' the same.  It's a minor fix, and I have a local test, but this sort of stuff needs a proper testsuite.

I am going to open an issue related to generating a decent set of tests for hardcaml features and pin the next version on getting this running.

**Comment by xguerin** [comment]
Excellent ! I'll give it a shot.

**Comment by xguerin** [comment]
**TL;DR** the generated code of my system is the same with or without the patch.

So as it turns out I don't have any shared instance proper. I use multiple instances of the same modules but as specialized functors, so in effect different instances.

**Comment by xguerin** [comment]
Wiring up the clock/reset was surprisingly easy. And the overhead is not even that much. Thanks again for the tip.

**Comment by andrewray** [comment]
I was thinking about this a bit more and I think you could also wrap submodules in a functor like;

```
module SomeSubmodule(Seq : Seq) = struct
    ...implementation using Seq.*** functions...
end
```

Then you wouldn't have to pass around clock everywhere and instantiate the Seq functor all the time.  You would only have to consider it when doing the actual instantiation of the submodule.

**Comment by xguerin** [comment]
Using a functor also crossed my mind. But I am still wondering how that would solve the signal hooking issue. Wouldn't it be still necessary to pass a `Sys.t` interface around ?

**Comment by andrewray** [comment]
It would still have to exist, but would only be actually needed when you instantiated the functor, which presumably would be at the point you generate the submodule.

I will need to have a play with an example to test this.  I also want to show how you can set up code so that it can generate either a hierarchical (for RTL) or flat design (for simulation) with a simple toggle.
